## Paso 1: Importar las Librerías Necesarias

In [6]:
# Importar librerías necesarias
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report, ConfusionMatrixDisplay, confusion_matrix
import matplotlib.pyplot as plt

## Paso 2: Cargar el DataFrame

In [16]:
# Cargar el archivo CSV con el DataFrame actualizado
df = pd.read_csv('encuestas_con_carrera_ideal.csv')

# Mostrar las primeras filas del DataFrame como una tabla
display(df.head())


Edad     Género País de nacimiento País de residencia  \
0    39  Masculino           Colombia             España   
1    24       Otro             México            Ecuador   
2    39       Otro              Chile             México   
3    44  Masculino            Ecuador             España   
4    39       Otro     Estados Unidos     Estados Unidos   

                                         Actividad 1  \
0  Explorar la inteligencia artificial y su aplic...   
1                        Investigar y analizar datos   
2                     Gestionar proyectos y recursos   
3                Desarrollar software o aplicaciones   
4  Diseñar contenido creativo (gráfico, audiovisual)   

                                         Actividad 2  \
0           Colaborar en equipos multidisciplinarios   
1  Escribir o crear contenido textual (artículos,...   
2  Explorar la inteligencia artificial y su aplic...   
3           Innovar en biotecnología o salud digital   
4          Consultar o asesorar sobre ciberseguridad   

                                         Actividad 3  \
0  Crear experiencias inmersivas en realidad virt...   
1  Explorar la inteligencia artificial y su aplic...   
2          Consultar o asesorar sobre ciberseguridad   
3                        Investigar y analizar datos   
4           Innovar en biotecnología o salud digital   

   Satisfacción Resolver problemas complejos  \
0                                          1   
1                                          1   
2                                          2   
3                                          1   
4                                          2   

   Satisfacción Colaborar en equipos  \
0                                  3   
1                                  5   
2                                  2   
3                                  2   
4                                  4   

   Satisfacción Investigar y analizar datos  ...  \
0                                         3  ...   
1                                         3  ...   
2                                         5  ...   
3                                         1  ...   
4                                         3  ...   

   Competencia Comunicación y oratoria                        Problema 1  \
0                                    1  Protección y privacidad de datos   
1                                    2          Educación y capacitación   
2                                    1         Justicia social y equidad   
3                                    4         Salud y bienestar público   
4                                    1          Educación y capacitación   

                         Problema 2  \
0         Justicia social y equidad   
1  Protección y privacidad de datos   
2         Justicia social y equidad   
3         Salud y bienestar público   
4          Educación y capacitación   

                                   Problema 3  \
0                   Salud y bienestar público   
1            Mejora de procesos empresariales   
2  Diseño de experiencias inmersivas en VR/AR   
3                    Educación y capacitación   
4            Protección y privacidad de datos   

                   Campo de trabajo preferido 1  \
0         Realidad virtual y realidad aumentada   
1                     Arte, diseño y multimedia   
2   Innovación sostenible y energías renovables   
3  Ingeniería (mecánica, civil, robótica, etc.)   
4                      Educación y capacitación   

           Campo de trabajo preferido 2     Campo de trabajo preferido 3  \
0  Ciberseguridad y protección de datos  Consultoría y análisis de datos   
1         Biotecnología y salud digital         Educación y capacitación   
2          Derecho y ciencias políticas  Consultoría y análisis de datos   
3              Educación y capacitación             Gestión de proyectos   
4         Biotecnología y salud digital        Arte, diseño y multimedia   

   Motivación económica (1-5)  Tipo d

## Paso 3: Preprocesamiento de los Datos
- Convertir variables categóricas a numéricas mediante la técnicas de codificación one-hot, para que el modelo de Machine Learning pueda trabajar con ellas.
- Definimos nuestras características (X) como las variables predictoras y el target (y) como la carrera ideal que queremos que el modelo prediga, preparando así los datos para el entrenamiento del modelo.

In [22]:
# Convertir variables categóricas a variables numéricas mediante codificación one-hot
df_encoded = pd.get_dummies(df.drop('carrera ideal', axis=1), drop_first=True)

# Definir X (características) e y (target)
X = df_encoded
y = df['carrera ideal']

# Verificar las dimensiones
print("Dimensiones de X:", X.shape)
print("Dimensiones de y:", y.shape)


Dimensiones de X: (300, 202)
Dimensiones de y: (300,)


In [28]:
from IPython.display import display

# Ver el DataFrame procesado (df_encoded)
display(df_encoded.head())


Edad  Satisfacción Resolver problemas complejos  \
0    39                                          1   
1    24                                          1   
2    39                                          2   
3    44                                          1   
4    39                                          2   

   Satisfacción Colaborar en equipos  \
0                                  3   
1                                  5   
2                                  2   
3                                  2   
4                                  4   

   Satisfacción Investigar y analizar datos  \
0                                         3   
1                                         3   
2                                         5   
3                                         1   
4                                         3   

   Satisfacción Diseñar contenido creativo  Satisfacción Enseñar o capacitar  \
0                                        5                                 4   
1                                        5                                 1   
2                                        3                                 2   
3                                        5                                 4   
4                                        3                                 2   

   Competencia Análisis de datos  Competencia Programación  \
0                              1                         1   
1                              5                         1   
2                              3                         2   
3                              4                         5   
4                              3                         4   

   Competencia Diseño gráfico y multimedia  Competencia Gestión de proyectos  \
0                                        4                                 5   
1                                        4                                 5   
2                                        2                                 3   
3                                        2                                 5   
4                                        5                                 4   

   ...  Campo de trabajo preferido 3_Derecho y ciencias políticas  \
0  ...                                              False           
1  ...                                              False           
2  ...                                              False           
3  ...                                              False           
4  ...                                              False           

   Campo de trabajo preferido 3_Educación y capacitación  \
0                                              False       
1                                               True       
2                                              False       
3                                              False       
4                                              False       

   Campo de trabajo preferido 3_Gestión de proyectos  \
0                                              False   
1                                              False   
2                                              False   
3                                               True   
4                                              False   

   Campo de trabajo preferido 3_Ingeniería (mecánica, civil, robótica, etc.)  \
0                                              False                           
1                                              False                           
2                                              False                           
3                                              False                           
4                                              False                           

   Campo de trabajo preferido 3_Innovación sostenible y energías renovables  \
0                                              False                          
1                                              False               

## Paso 4: Dividir los Datos en Conjuntos de Entrenamiento y Prueba

In [37]:
from sklearn.model_selection import train_test_split

# Dividir los datos en conjunto de entrenamiento (70%) y conjunto de prueba (30%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

print("Dimensiones del conjunto de entrenamiento:")
print("X_train:", X_train.shape, "y_train:", y_train.shape)
print("Dimensiones del conjunto de prueba:")
print("X_test:", X_test.shape, "y_test:", y_test.shape)


Dimensiones del conjunto de entrenamiento:
X_train: (210, 202) y_train: (210,)
Dimensiones del conjunto de prueba:
X_test: (90, 202) y_test: (90,)


## Paso 5: Entrenar el Modelo de Árbol de Decisión

In [42]:
# Crear y entrenar el modelo de árbol de decisión
clf = DecisionTreeClassifier(random_state=42)
clf.fit(X_train, y_train)

# Predecir en el conjunto de prueba
y_pred = clf.predict(X_test)

# Evaluar el modelo
print("Exactitud del modelo:", accuracy_score(y_test, y_pred))
print("Reporte de clasificación:")
print(classification_report(y_test, y_pred))


Exactitud del modelo: 0.05555555555555555
Reporte de clasificación:
                                             precision    recall  f1-score   support

                          Abogado Ambiental       0.00      0.00      0.00         1
                            Abogado Digital       0.00      0.00      0.00         1
                        Abogado Tradicional       0.00      0.00      0.00         3
                  Administrador de Empresas       0.00      0.00      0.00         1
                          Analista de Datos       0.00      0.00      0.00         2
      Analista de Desarrollo Organizacional       0.00      0.00      0.00         2
    Analista de Experiencia de Usuario (UX)       0.00      0.00      0.00         1
              Analista de Innovación Social       0.00      0.00      0.00         1
                                 Arquitecto       0.00      0.00      0.00         0
           Arquitecto de Realidad Aumentada       0.00      0.00      0.00       

/opt/anaconda3/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.12/site-packages/sklearn/metrics/

## primeros resultados:
1. Bajo Desempeño General
La exactitud del modelo es extremadamente baja (5.56%). Esto indica que el modelo de árbol de decisión apenas está haciendo predicciones correctas y no es capaz de aprender patrones útiles a partir de los datos proporcionados.

2. Clase Desbalanceada y Alta Variedad de Clases
El target 'carrera ideal' parece tener un gran número de clases diferentes (carreras específicas), muchas de las cuales tienen muy pocos ejemplos en los datos de entrenamiento. Esto crea un problema de clase desbalanceada, donde el modelo no tiene suficientes datos para aprender correctamente cómo clasificar cada carrera.
Las clases con más datos muestran una ligera mejora en las métricas, pero la mayoría de las carreras están muy poco representadas, lo que dificulta el aprendizaje del modelo.

3. Necesidad de Mejorar la Calidad de los Datos y el Modelo
El alto número de características (202) y la posible codificación de variables categóricas podrían estar generando ruido en el modelo, lo que hace que sea difícil identificar patrones relevantes.
- Dado que hay muchas clases, una opción sería agrupar las carreras en categorías más generales, lo que podría ayudar al modelo a encontrar patrones más generales antes de hacer predicciones específicas.
-Es probable que el tamaño del conjunto de datos (300 ejemplos) sea insuficiente para capturar las complejidades de un problema tan amplio y variado.


## Vamos a probar con el modelo Random Forest, para mejorar el rendimiento:

In [55]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, ConfusionMatrixDisplay, confusion_matrix
import matplotlib.pyplot as plt
import pandas as pd


In [57]:
# Crear y entrenar el modelo de Random Forest
rf_clf = RandomForestClassifier(random_state=42, n_estimators=100)  # Puedes ajustar n_estimators para cambiar el número de árboles
rf_clf.fit(X_train, y_train)

# Predecir en el conjunto de prueba
y_pred = rf_clf.predict(X_test)

# Evaluar el modelo
print("Exactitud del modelo:", accuracy_score(y_test, y_pred))
print("Reporte de clasificación:")
print(classification_report(y_test, y_pred))


Exactitud del modelo: 0.06666666666666667
Reporte de clasificación:
                                             precision    recall  f1-score   support

                          Abogado Ambiental       0.00      0.00      0.00         1
                            Abogado Digital       0.00      0.00      0.00         1
                        Abogado Tradicional       0.00      0.00      0.00         3
                  Administrador de Empresas       0.00      0.00      0.00         1
                          Analista de Datos       0.00      0.00      0.00         2
      Analista de Desarrollo Organizacional       0.00      0.00      0.00         2
    Analista de Experiencia de Usuario (UX)       0.00      0.00      0.00         1
              Analista de Innovación Social       0.00      0.00      0.00         1
                                 Arquitecto       0.00      0.00      0.00         0
           Arquitecto de Realidad Aumentada       0.00      0.00      0.00       

/opt/anaconda3/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.12/site-packages/sklearn/metrics/

## Resultados:

Bajo Rendimiento del Modelo:

- La precisión general del modelo sigue siendo baja, lo que sugiere que las características disponibles no son suficientemente informativas para predecir con precisión la carrera ideal de los usuarios.

- La mayoría de las clases tienen valores cercanos a cero en precision, recall y f1-score, lo que indica que el modelo no puede distinguir entre muchas de las posibles carreras.

- Similar al modelo anterior, el target 'carrera ideal' tiene muchas clases únicas (carreras específicas) con muy pocos ejemplos para cada una, lo que dificulta el aprendizaje del modelo. Las clases que tienen pocos ejemplos no están siendo bien representadas en el entrenamiento.
- Posible Complejidad del Problema: Dado que el problema implica predecir una gran cantidad de carreras basadas en muchas características, el modelo puede estar sobreajustándose o simplemente no ser capaz de encontrar patrones claros debido a la complejidad y diversidad de las carreras.

## Posibles solusiones:

- Agrupación de Carreras en Categorías Más Generales
- Aumentar el Conjunto de Datos
- Explorar Otros Modelos: Modelos de Ensemblado como Gradient Boosting o XGBoost pueden ofrecer mejoras o Modelos más complejos como Redes Neuronales podrían ser útiles con suficientes datos.
- Validación y Cross-Validation